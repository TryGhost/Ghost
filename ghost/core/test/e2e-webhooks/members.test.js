const {agentProvider, mockManager, fixtureManager, matchers} = require('../utils/e2e-framework');
const {anyGhostAgent, anyObjectId, anyISODateTime, anyUuid, anyContentVersion, anyNumber} = matchers;

const buildNewsletterSnapshot = () => {
    const newsLetterSnapshot = {
        id: anyObjectId
    };

    return newsLetterSnapshot;
};

const buildMemberSnapshot = () => {
    const memberSnapshot = {
        id: anyObjectId,
        uuid: anyUuid,
        created_at: anyISODateTime,
        updated_at: anyISODateTime,
        newsletters: new Array(1).fill(buildNewsletterSnapshot())
    };

    return memberSnapshot;
};

describe('member.* events', function () {
    let adminAPIAgent;
    let webhookMockReceiver;

    before(async function () {
        adminAPIAgent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('integrations');
        await adminAPIAgent.loginAsOwner();
    });

    beforeEach(function () {
        webhookMockReceiver = mockManager.mockWebhookRequests();
    });

    afterEach(function () {
        mockManager.restore();
    });

    it('member.added event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/member-added/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'member.added',
            url: webhookURL
        });

        await adminAPIAgent
            .post('members/')
            .body({
                members: [{
                    name: 'Test Member',
                    email: 'testemail@example.com',
                    note: 'test note'
                }]
            })
            .expectStatus(201);

        await webhookMockReceiver.receivedRequest();

        webhookMockReceiver
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                'content-length': anyNumber,
                'user-agent': anyGhostAgent
            })
            .matchBodySnapshot({
                member: {
                    current: buildMemberSnapshot()
                }
            });
    });

    it('member.deleted event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/member-deleted/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'member.deleted',
            url: webhookURL
        });

        const res = await adminAPIAgent
            .post('members/')
            .body({
                members: [{
                    name: 'Test Member2',
                    email: 'testemail2@example.com',
                    note: 'test note2'
                }]
            })
            .expectStatus(201);

        const id = res.body.members[0].id;

        await adminAPIAgent
            .delete('members/' + id)
            .expectStatus(204);

        await webhookMockReceiver.receivedRequest();

        webhookMockReceiver
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                'content-length': anyNumber,
                'user-agent': anyGhostAgent
            })
            .matchBodySnapshot({
                member: {
                    current: {},
                    previous: buildMemberSnapshot()
                }
            });
    });

    it('member.edited event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/member-edited/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'member.edited',
            url: webhookURL
        });

        const res = await adminAPIAgent
            .post('members/')
            .body({
                members: [{
                    name: 'Test Member3',
                    email: 'testemail3@example.com',
                    note: 'test note3'
                }]
            })
            .expectStatus(201);

        const id = res.body.members[0].id;

        await adminAPIAgent
            .put('members/' + id)
            .body({
                members: [{name: 'Ghost'}]
            })
            .expectStatus(200);

        await webhookMockReceiver.receivedRequest();

        webhookMockReceiver
            .matchHeaderSnapshot({
                'content-version': anyContentVersion,
                'content-length': anyNumber,
                'user-agent': anyGhostAgent
            })
            .matchBodySnapshot({
                member: {
                    current: buildMemberSnapshot(),
                    previous: {
                        updated_at: anyISODateTime
                    }
                }
            });
    });
});
