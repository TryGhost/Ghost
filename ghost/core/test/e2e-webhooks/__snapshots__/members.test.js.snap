// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`member.* events member.added event is triggered 1: [headers] 1`] = `
Object {
  "accept-encoding": "gzip, deflate, br",
  "content-length": Any<Number>,
  "content-type": "application/json",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "user-agent": StringMatching /Ghost\\\\/\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\\\s\\\\\\(https:\\\\/\\\\/github\\.com\\\\/TryGhost\\\\/Ghost\\\\\\)/,
}
`;

exports[`member.* events member.added event is triggered 2: [body] 1`] = `
Object {
  "member": Object {
    "current": Object {
      "avatar_image": null,
      "comped": false,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "email": "testemail@example.com",
      "email_count": 0,
      "email_open_rate": null,
      "email_opened_count": 0,
      "geolocation": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "last_seen_at": null,
      "name": "Test Member",
      "newsletters": Array [
        Object {
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "name": "Default Newsletter",
          "status": "active",
        },
      ],
      "note": "test note",
      "status": "free",
      "subscribed": true,
      "subscriptions": Array [],
      "tiers": Array [],
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
    },
    "previous": Object {},
  },
}
`;

exports[`member.* events member.deleted event is triggered 1: [headers] 1`] = `
Object {
  "accept-encoding": "gzip, deflate, br",
  "content-length": Any<Number>,
  "content-type": "application/json",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "user-agent": StringMatching /Ghost\\\\/\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\\\s\\\\\\(https:\\\\/\\\\/github\\.com\\\\/TryGhost\\\\/Ghost\\\\\\)/,
}
`;

exports[`member.* events member.deleted event is triggered 2: [body] 1`] = `
Object {
  "member": Object {
    "current": Object {},
    "previous": Object {
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "email": "testemail2@example.com",
      "email_count": 0,
      "email_open_rate": null,
      "email_opened_count": 0,
      "geolocation": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "last_seen_at": null,
      "name": "Test Member2",
      "newsletters": Array [
        Object {
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "name": "Default Newsletter",
          "status": "active",
        },
      ],
      "note": "test note2",
      "status": "free",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
    },
  },
}
`;

exports[`member.* events member.edited event is triggered 1: [headers] 1`] = `
Object {
  "accept-encoding": "gzip, deflate, br",
  "content-length": Any<Number>,
  "content-type": "application/json",
  "content-version": StringMatching /v\\\\d\\+\\\\\\.\\\\d\\+/,
  "user-agent": StringMatching /Ghost\\\\/\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\\\s\\\\\\(https:\\\\/\\\\/github\\.com\\\\/TryGhost\\\\/Ghost\\\\\\)/,
}
`;

exports[`member.* events member.edited event is triggered 2: [body] 1`] = `
Object {
  "member": Object {
    "current": Object {
      "avatar_image": null,
      "comped": false,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "email": "testemail3@example.com",
      "email_count": 0,
      "email_open_rate": null,
      "email_opened_count": 0,
      "geolocation": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "labels": Array [],
      "last_seen_at": null,
      "name": "Ghost",
      "newsletters": Array [
        Object {
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "name": "Default Newsletter",
          "status": "active",
        },
      ],
      "note": "test note3",
      "status": "free",
      "subscribed": true,
      "subscriptions": Array [],
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
    },
    "previous": Object {
      "name": "Test Member3",
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
    },
  },
}
`;
